src/garage/tf/algos/rl2_npo.py

```diff
@@ -0,0 +1,733 @@
+                action_flat = flatten_batch(action_var, name='action_flat')
+                reward_flat = flatten_batch(reward_var, name='reward_flat')
+                valid_flat = flatten_batch(valid_var, name='valid_flat')
+                policy_state_info_vars_flat = flatten_batch_dict(
+                    policy_state_info_vars, name='policy_state_info_vars_flat')
+                policy_old_dist_info_vars_flat = flatten_batch_dict(
+                    policy_old_dist_info_vars,
+                    name='policy_old_dist_info_vars_flat')
+
+            # valid view
+            with tf.name_scope('valid'):
+                action_valid = filter_valids(action_flat,
+                                             valid_flat,
+                                             name='action_valid')
+                policy_state_info_vars_valid = filter_valids_dict(
+                    policy_state_info_vars_flat,
+                    valid_flat,
+                    name='policy_state_info_vars_valid')
+                policy_old_dist_info_vars_valid = filter_valids_dict(
+                    policy_old_dist_info_vars_flat,
+                    valid_flat,
+                    name='policy_old_dist_info_vars_valid')
+
+        # policy loss and optimizer inputs
+        pol_flat = graph_inputs(
+            'PolicyLossInputsFlat',
+            obs_var=obs_flat,
+            action_var=action_flat,
+            reward_var=reward_flat,
+            valid_var=valid_flat,
+            policy_state_info_vars=policy_state_info_vars_flat,
+            policy_old_dist_info_vars=policy_old_dist_info_vars_flat,
+        )
+        pol_valid = graph_inputs(
+            'PolicyLossInputsValid',
+            action_var=action_valid,
+            policy_state_info_vars=policy_state_info_vars_valid,
+            policy_old_dist_info_vars=policy_old_dist_info_vars_valid,
+        )
+        policy_loss_inputs = graph_inputs(
+            'PolicyLossInputs',
+            obs_var=obs_var,
+            action_var=action_var,
+            reward_var=reward_var,
+            baseline_var=baseline_var,
+            valid_var=valid_var,
+            length_var=length_var,
+            policy_state_info_vars=policy_state_info_vars,
+            policy_old_dist_info_vars=policy_old_dist_info_vars,
+            flat=pol_flat,
+            valid=pol_valid,
+        )
+        policy_opt_inputs = graph_inputs(
+            'PolicyOptInputs',
+            obs_var=obs_var,
+            action_var=action_var,
+            reward_var=reward_var,
+            baseline_var=baseline_var,
+            valid_var=valid_var,
+            length_var=length_var,
+            policy_state_info_vars_list=policy_state_info_vars_list,
+            policy_old_dist_info_vars_list=policy_old_dist_info_vars_list,
+        )
+
+        return policy_loss_inputs, policy_opt_inputs
+
+    # pylint: disable=too-many-branches, too-many-statements
+    def _build_policy_loss(self, i):
```

Comment:
This is way too much copied code. There are definitely better ways to implement RL2 without copying all of this.

Reply:
After discussing this and reviewing the code carefully, I think Ryan and I understand a lot better what's going on here.
We also re-read part of the original RL^2 paper. We discovered a few things:
 - Originally, RL^2 was implemented with all tasks using the same RNN / GRU baseline. ProMP doesn't do this.
 - According to the paper, advantages are computed across the "trial" a.k.a. all paths in the rollout. This means that you shouldn't need `compute_advantages_meta_learn` or `discounted_returns_meta_learn`, since you can just call the original methods on the concatenated paths.

As for the implementation, we think you can avoid copying by doing the following:
 - Modify the NPO implementation by refactoring out all of the parts you need to change. For example, you'll probably need to create an empty `fit_baseline_early()` method in NPO. Then, create a subclass of NPO (maybe called `_NPO_RL2` which is used as an `_inner_algo` in your `RL2` class. This subclass can over-ride all of the unique behavior (such as implementing the `fit_baseline_early()` method), without needing to copy any code.

Sorry if my prior advice on inheriting from `BatchPolopt` lead you to doing it this way. Hopefully this plan makes sense?
Let me know if you have any questions, or find yourself copying code. It should be possible to avoid. Thanks.